File: tests/unit/m3u-parser.test.ts
import { describe, it, expect } from 'vitest';
import { parseM3u } from '../../src/main/services/m3u-parser';

describe('M3U Parser', () => {
  it('parses a basic M3U playlist', () => {
    const content = `#EXTM3U
#EXTINF:-1 tvg-id="ch1" tvg-name="Channel 1" tvg-logo="http://logo.com/1.png" group-title="News",Channel One HD
http://stream.example.com/ch1
#EXTINF:-1 tvg-id="ch2" tvg-name="Channel 2" tvg-logo="http://logo.com/2.png" group-title="Sports",ESPN Live
http://stream.example.com/ch2`;

    const entries = parseM3u(content);
    expect(entries).toHaveLength(2);

    expect(entries[0].title).toBe('Channel One HD');
    expect(entries[0].tvgId).toBe('ch1');
    expect(entries[0].tvgName).toBe('Channel 1');
    expect(entries[0].tvgLogo).toBe('http://logo.com/1.png');
    expect(entries[0].groupTitle).toBe('News');
    expect(entries[0].streamUrl).toBe('http://stream.example.com/ch1');
    expect(entries[0].duration).toBe(-1);

    expect(entries[1].title).toBe('ESPN Live');
    expect(entries[1].groupTitle).toBe('Sports');
  });

  it('handles BOM marker', () => {
    const content = '\uFEFF#EXTM3U\n#EXTINF:-1,Test Channel\nhttp://example.com/stream';
    const entries = parseM3u(content);
    expect(entries).toHaveLength(1);
    expect(entries[0].title).toBe('Test Channel');
  });

  it('handles Windows line endings (CRLF)', () => {
    const content = '#EXTM3U\r\n#EXTINF:-1,Channel A\r\nhttp://a.com/1\r\n#EXTINF:-1,Channel B\r\nhttp://b.com/2\r\n';
    const entries = parseM3u(content);
    expect(entries).toHaveLength(2);
    expect(entries[0].title).toBe('Channel A');
    expect(entries[1].title).toBe('Channel B');
  });

  it('handles old Mac line endings (CR only)', () => {
    const content = '#EXTM3U\r#EXTINF:-1,Channel X\rhttp://x.com/stream';
    const entries = parseM3u(content);
    expect(entries).toHaveLength(1);
    expect(entries[0].title).toBe('Channel X');
  });

  it('skips empty lines', () => {
    const content = `#EXTM3U

#EXTINF:-1,Channel 1

http://example.com/1

#EXTINF:-1,Channel 2

http://example.com/2
`;
    const entries = parseM3u(content);
    expect(entries).toHaveLength(2);
  });

  it('handles malformed entries — URL without EXTINF', () => {
    const content = `#EXTM3U
http://example.com/stream1
#EXTINF:-1,Proper Channel
http://example.com/stream2`;

    const entries = parseM3u(content);
    expect(entries).toHaveLength(2);
    // First entry has a generated title from URL
    expect(entries[0].streamUrl).toBe('http://example.com/stream1');
    // Second entry is properly parsed
    expect(entries[1].title).toBe('Proper Channel');
  });

  it('extracts duration correctly', () => {
    const content = `#EXTM3U
#EXTINF:3600,Movie Title
http://example.com/movie
#EXTINF:-1,Live Channel
http://example.com/live
#EXTINF:0,Unknown Duration
http://example.com/unknown`;

    const entries = parseM3u(content);
    expect(entries[0].duration).toBe(3600);
    expect(entries[1].duration).toBe(-1);
    expect(entries[2].duration).toBe(0);
  });

  it('handles EXTINF with no attributes — just duration and title', () => {
    const content = `#EXTM3U
#EXTINF:-1,Simple Channel
http://example.com/simple`;

    const entries = parseM3u(content);
    expect(entries).toHaveLength(1);
    expect(entries[0].title).toBe('Simple Channel');
    expect(entries[0].groupTitle).toBe('');
    expect(entries[0].tvgId).toBe('');
  });

  it('handles attributes with single quotes', () => {
    const content = `#EXTM3U
#EXTINF:-1 tvg-id='abc' group-title='Entertainment',Show Name
http://example.com/show`;

    const entries = parseM3u(content);
    expect(entries[0].tvgId).toBe('abc');
    expect(entries[0].groupTitle).toBe('Entertainment');
  });

  it('handles real-world provider titles with noise', () => {
    const content = `#EXTM3U
#EXTINF:-1 tvg-id="us.cnn" tvg-logo="http://cdn.logo/cnn.png" group-title="US | News",US: CNN HD [MULTI]
http://provider.com/live/us/cnn
#EXTINF:-1 tvg-id="" tvg-logo="" group-title="VOD | Movies",The Matrix (1999) [4K]
http://provider.com/movie/12345.mp4
#EXTINF:-1 tvg-id="" group-title="Series | Drama",Breaking Bad S01E01
http://provider.com/series/bb/s01e01.mp4`;

    const entries = parseM3u(content);
    expect(entries).toHaveLength(3);
    expect(entries[0].title).toBe('US: CNN HD [MULTI]');
    expect(entries[0].groupTitle).toBe('US | News');
    expect(entries[1].title).toBe('The Matrix (1999) [4K]');
    expect(entries[2].title).toBe('Breaking Bad S01E01');
  });

  it('handles empty playlist', () => {
    const content = '#EXTM3U\n';
    const entries = parseM3u(content);
    expect(entries).toHaveLength(0);
  });

  it('handles playlist with only #EXTM3U header', () => {
    const content = '#EXTM3U';
    const entries = parseM3u(content);
    expect(entries).toHaveLength(0);
  });

  it('skips other directives like #EXTVLCOPT', () => {
    const content = `#EXTM3U
#EXTINF:-1,Channel 1
#EXTVLCOPT:http-user-agent=Mozilla
#EXTVLCOPT:http-referrer=http://ref.com
http://example.com/ch1`;

    const entries = parseM3u(content);
    expect(entries).toHaveLength(1);
    expect(entries[0].title).toBe('Channel 1');
    expect(entries[0].streamUrl).toBe('http://example.com/ch1');
  });

  it('handles group-title with special characters', () => {
    const content = `#EXTM3U
#EXTINF:-1 group-title="US | News & Politics (HD)",Channel
http://example.com/ch`;

    const entries = parseM3u(content);
    expect(entries[0].groupTitle).toBe('US | News & Politics (HD)');
  });

  it('handles large number of entries efficiently', () => {
    const lines = ['#EXTM3U'];
    for (let i = 0; i < 10000; i++) {
      lines.push(`#EXTINF:-1 group-title="Group ${i % 10}",Channel ${i}`);
      lines.push(`http://example.com/ch${i}`);
    }
    const content = lines.join('\n');

    const start = Date.now();
    const entries = parseM3u(content);
    const elapsed = Date.now() - start;

    expect(entries).toHaveLength(10000);
    expect(elapsed).toBeLessThan(5000); // Should parse 10K entries in under 5 seconds
  });
});

File: packages/shared/src/commonTest/kotlin/com/yancotv/shared/parsers/M3uParserTest.kt
package com.yancotv.shared.parsers

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Mirrors `tests/unit/m3u-parser.test.ts`. When the TS test file changes
 * ports of added cases land here in the same commit.
 */
class M3uParserTest {
    @Test
    fun parsesBasicPlaylist() {
        val content = """#EXTM3U
#EXTINF:-1 tvg-id="ch1" tvg-name="Channel 1" tvg-logo="http://logo.com/1.png" group-title="News",Channel One HD
http://stream.example.com/ch1
#EXTINF:-1 tvg-id="ch2" tvg-name="Channel 2" tvg-logo="http://logo.com/2.png" group-title="Sports",ESPN Live
http://stream.example.com/ch2"""

        val (entries, _) = parseM3u(content)
        assertEquals(2, entries.size)

        assertEquals("Channel One HD", entries[0].title)
        assertEquals("ch1", entries[0].tvgId)
        assertEquals("Channel 1", entries[0].tvgName)
        assertEquals("http://logo.com/1.png", entries[0].tvgLogo)
        assertEquals("News", entries[0].groupTitle)
        assertEquals("http://stream.example.com/ch1", entries[0].streamUrl)
        assertEquals(-1.0, entries[0].duration)

        assertEquals("ESPN Live", entries[1].title)
        assertEquals("Sports", entries[1].groupTitle)
    }

    @Test
    fun handlesBomMarker() {
        val content = "\uFEFF#EXTM3U\n#EXTINF:-1,Test Channel\nhttp://example.com/stream"
        val (entries, _) = parseM3u(content)
        assertEquals(1, entries.size)
        assertEquals("Test Channel", entries[0].title)
    }

    @Test
    fun handlesCrlfLineEndings() {
        val content = "#EXTM3U\r\n#EXTINF:-1,Channel A\r\nhttp://a.com/1\r\n#EXTINF:-1,Channel B\r\nhttp://b.com/2\r\n"
        val (entries, _) = parseM3u(content)
        assertEquals(2, entries.size)
        assertEquals("Channel A", entries[0].title)
        assertEquals("Channel B", entries[1].title)
    }

    @Test
    fun handlesCrOnlyLineEndings() {
        val content = "#EXTM3U\r#EXTINF:-1,Channel X\rhttp://x.com/stream"
        val (entries, _) = parseM3u(content)
        assertEquals(1, entries.size)
        assertEquals("Channel X", entries[0].title)
    }

    @Test
    fun skipsEmptyLines() {
        val content = """#EXTM3U

#EXTINF:-1,Channel 1

http://example.com/1

#EXTINF:-1,Channel 2

http://example.com/2
"""
        val (entries, _) = parseM3u(content)
        assertEquals(2, entries.size)
    }

    @Test
    fun handlesUrlWithoutExtinf() {
        val content = """#EXTM3U
http://example.com/stream1
#EXTINF:-1,Proper Channel
http://example.com/stream2"""

        val (entries, _) = parseM3u(content)
        assertEquals(2, entries.size)
        assertEquals("http://example.com/stream1", entries[0].streamUrl)
        assertEquals("Proper Channel", entries[1].title)
    }

    @Test
    fun extractsDuration() {
        val content = """#EXTM3U
#EXTINF:3600,Movie Title
http://example.com/movie
#EXTINF:-1,Live Channel
http://example.com/live
#EXTINF:0,Unknown Duration
http://example.com/unknown"""

        val (entries, _) = parseM3u(content)
        assertEquals(3600.0, entries[0].duration)
        assertEquals(-1.0, entries[1].duration)
        assertEquals(0.0, entries[2].duration)
    }

    @Test
    fun handlesExtinfNoAttributes() {
        val content = """#EXTM3U
#EXTINF:-1,Simple Channel
http://example.com/simple"""

        val (entries, _) = parseM3u(content)
        assertEquals(1, entries.size)
        assertEquals("Simple Channel", entries[0].title)
        assertEquals("", entries[0].groupTitle)
        assertEquals("", entries[0].tvgId)
    }

    @Test
    fun handlesSingleQuotedAttributes() {
        val content = """#EXTM3U
#EXTINF:-1 tvg-id='abc' group-title='Entertainment',Show Name
http://example.com/show"""

        val (entries, _) = parseM3u(content)
        assertEquals("abc", entries[0].tvgId)
        assertEquals("Entertainment", entries[0].groupTitle)
    }

    @Test
    fun handlesRealWorldTitlesWithNoise() {
        val content = """#EXTM3U
#EXTINF:-1 tvg-id="us.cnn" tvg-logo="http://cdn.logo/cnn.png" group-title="US | News",US: CNN HD [MULTI]
http://provider.com/live/us/cnn
#EXTINF:-1 tvg-id="" tvg-logo="" group-title="VOD | Movies",The Matrix (1999) [4K]
http://provider.com/movie/12345.mp4
#EXTINF:-1 tvg-id="" group-title="Series | Drama",Breaking Bad S01E01
http://provider.com/series/bb/s01e01.mp4"""

        val (entries, _) = parseM3u(content)
        assertEquals(3, entries.size)
        assertEquals("US: CNN HD [MULTI]", entries[0].title)
        assertEquals("US | News", entries[0].groupTitle)
        assertEquals("The Matrix (1999) [4K]", entries[1].title)
        assertEquals("Breaking Bad S01E01", entries[2].title)
    }

    @Test
    fun handlesEmptyPlaylist() {
        val content = "#EXTM3U\n"
        val (entries, _) = parseM3u(content)
        assertEquals(0, entries.size)
    }

    @Test
    fun handlesHeaderOnly() {
        val content = "#EXTM3U"
        val (entries, _) = parseM3u(content)
        assertEquals(0, entries.size)
    }

    @Test
    fun skipsOtherDirectives() {
        val content = """#EXTM3U
#EXTINF:-1,Channel 1
#EXTVLCOPT:http-user-agent=Mozilla
#EXTVLCOPT:http-referrer=http://ref.com
http://example.com/ch1"""

        val (entries, _) = parseM3u(content)
        assertEquals(1, entries.size)
        assertEquals("Channel 1", entries[0].title)
        assertEquals("http://example.com/ch1", entries[0].streamUrl)
    }

    @Test
    fun handlesSpecialCharsInGroupTitle() {
        val content = """#EXTM3U
#EXTINF:-1 group-title="US | News & Politics (HD)",Channel
http://example.com/ch"""

        val (entries, _) = parseM3u(content)
        assertEquals("US | News & Politics (HD)", entries[0].groupTitle)
    }

    @Test
    fun extractsUrlTvgEpgUrl() {
        val content = """#EXTM3U url-tvg="http://epg.example.com/guide.xml.gz"
#EXTINF:-1 tvg-id="ch1",Channel 1
http://stream.example.com/ch1"""

        val result = parseM3u(content)
        assertEquals("http://epg.example.com/guide.xml.gz", result.epgUrl)
        assertEquals(1, result.entries.size)
    }

    @Test
    fun extractsXTvgUrlEpgUrl() {
        val content = """#EXTM3U x-tvg-url="http://epg.example.com/alt.xml"
#EXTINF:-1,Channel 1
http://stream.example.com/ch1"""

        val result = parseM3u(content)
        assertEquals("http://epg.example.com/alt.xml", result.epgUrl)
    }

    @Test
    fun returnsNullEpgUrlWhenAbsent() {
        val content = """#EXTM3U
#EXTINF:-1,Channel 1
http://stream.example.com/ch1"""

        val result = parseM3u(content)
        assertNull(result.epgUrl)
    }

    @Test
    fun handlesLargePlaylist() {
        val lines = StringBuilder("#EXTM3U")
        for (i in 0 until 10000) {
            lines.append("\n#EXTINF:-1 group-title=\"Group ${i % 10}\",Channel $i")
            lines.append("\nhttp://example.com/ch$i")
        }

        val result = parseM3u(lines.toString())
        assertEquals(10000, result.entries.size)
        assertTrue(result.entries[0].title == "Channel 0")
        assertTrue(result.entries.last().title == "Channel 9999")
    }
}
